node/network/bitfield-distribution/src/lib.rs

```diff
@@ -0,0 +1,684 @@
+	let validator = if let Some(validator) = validator_set.get(validator_index) {
+		validator.clone()
+	} else {
+		return modify_reputation(ctx, origin, COST_VALIDATOR_INDEX_INVALID).await;
+	};
+
+	if message
+		.signed_availability
+		.check_signature(&signing_context, &validator)
+		.is_ok()
+	{
+		let one_per_validator = &mut (job_data.one_per_validator);
+		// only relay_message a message of a validator once
+		if one_per_validator.get(&validator).is_some() {
+			trace!(target: "bitd", "Alrady received a message for validator at index {}", validator_index);
+			return Ok(());
+		}
+		one_per_validator.insert(validator.clone(), message.clone());
+
+		relay_message(ctx, job_data, &mut tracker.peer_views, validator, message).await
+	} else {
+		return modify_reputation(ctx, origin, COST_SIGNATURE_INVALID).await;
+	}
+}
+
+/// Deal with network bridge updates and track what needs to be tracked
+/// which depends on the message type received.
+async fn handle_network_msg<Context>(
+	ctx: &mut Context,
+	tracker: &mut Tracker,
+	bridge_message: NetworkBridgeEvent,
+) -> SubsystemResult<()>
+where
+	Context: SubsystemContext<Message = BitfieldDistributionMessage>,
+{
+	match bridge_message {
+		NetworkBridgeEvent::PeerConnected(peerid, _role) => {
+			// insert if none already present
+			tracker.peer_views.entry(peerid).or_insert(View::default());
+		}
+		NetworkBridgeEvent::PeerDisconnected(peerid) => {
+			// get rid of superfluous data
+			tracker.peer_views.remove(&peerid);
+		}
+		NetworkBridgeEvent::PeerViewChange(peerid, view) => {
+			catch_up_messages(ctx, tracker, peerid, view).await?;
+		}
+		NetworkBridgeEvent::OurViewChange(view) => {
+			let old_view = std::mem::replace(&mut (tracker.view), view);
+
+			for new in tracker.view.difference(&old_view) {
+				if !tracker.per_relay_parent.contains_key(&new) {
+					warn!(target: "bitd", "Our view contains {} but the overseer never told use we should work on this", &new);
+				}
+			}
+		}
+		NetworkBridgeEvent::PeerMessage(remote, bytes) => {
+			if let Ok(gossiped_bitfield) = BitfieldGossipMessage::decode(&mut (bytes.as_slice())) {
+				trace!(target: "bitd", "Received bitfield gossip from peer {:?}", &remote);
+				process_incoming_peer_message(ctx, tracker, remote, gossiped_bitfield).await?;
+			} else {
+				return modify_reputation(ctx, remote, COST_MESSAGE_NOT_DECODABLE).await;
+			}
+		}
+	}
+	Ok(())
+}
+
+// Send the difference between two views which were not sent
+// to that particular peer.
+async fn catch_up_messages<Context>(
+	ctx: &mut Context,
+	tracker: &mut Tracker,
+	origin: PeerId,
+	view: View,
+) -> SubsystemResult<()>
+where
+	Context: SubsystemContext<Message = BitfieldDistributionMessage>,
+{
+	use std::collections::hash_map::Entry;
+	let current = tracker.peer_views.entry(origin.clone()).or_default();
+
+	let delta_vec: Vec<Hash> = (*current).difference(&view).cloned().collect();
+
+	*current = view;
```

Comment:
Based on the name and documentation of the function, I wouldn't have expected a bookkeeping update. However, I do in general expect this bookkeeping update. Recommend a rename of function to `handle_peer_view_change`.